File: android/src/main/java/com/lightningkite/khrysalistemplate/vg/BleScanDemoVG.shared.kt
package com.lightningkite.khrysalistemplate.vg

import android.widget.*
import android.view.*
import com.lightningkite.khrysalis.*
import com.lightningkite.khrysalis.bluetooth.Ble
import com.lightningkite.khrysalis.bluetooth.BleDeviceInfo
import com.lightningkite.khrysalis.bluetooth.BleScanResult
import com.lightningkite.khrysalis.lifecycle.*
import com.lightningkite.khrysalis.views.*
import com.lightningkite.khrysalis.observables.*
import com.lightningkite.khrysalis.observables.binding.*
import com.lightningkite.khrysalis.rx.removed
import com.lightningkite.khrysalis.rx.until
import com.lightningkite.khrysalistemplate.R
import com.lightningkite.khrysalistemplate.layouts.*
import io.reactivex.rxkotlin.subscribeBy

@Suppress("NAME_SHADOWING")
class BleScanDemoVG(
    val stack: ObservableStack<ViewGenerator>
) : ViewGenerator() {

    override val title: String get() = "BleScanDemo"

    val devices: MutableObservableProperty<Map<String, BleScanResult>> = StandardObservableProperty(mapOf())

    override fun generate(dependency: ViewDependency): View {
        val xml = BleScanDemoXml()
        val view = xml.setup(dependency)

        xml.button.onClick(captureWeak(xml.button){ button ->
            Ble.scan(viewDependency = dependency)
                .subscribeBy { result ->
                    this.devices.value = this.devices.value + Pair(result.info.id, result)
                }
                .until(button.removed)
        })

        xml.host.onClick {
            this.stack.push(BleHostDemoVG())
        }

        xml.items.bind(
            data = devices.map { it -> it.values.sortedByDescending { it -> it.rssi } },
            defaultValue = BleScanResult(BleDeviceInfo("", ""), -1000),
            makeView = label@ { obs ->
                val cellXml = ComponentBleDeviceXml()
                val cellView = cellXml.setup(dependency)
                cellXml.deviceName.bindString(obs.map { it -> it.info.name ?: "No Name"})
                cellXml.deviceId.bindString(obs.map { it -> it.info.id })
                cellXml.rssi.bindString(obs.map { it -> it.rssi.toString() })
                cellXml.connect.onClick {  }
                return@label cellView
            }
        )

        return view
    }

}

File: ios/KhrysalisTemplate/com/lightningkite/khrysalistemplate/vg/BleScanDemoVG.shared.swift
//Package: com.lightningkite.khrysalistemplate.vg
//Converted using Khrysalis2

import Foundation
import Khrysalis
import RxSwift
import RxRelay
import KhrysalisMaps
import KhrysalisBluetooth



public class BleScanDemoVG: ViewGenerator {
    
    public var stack: ObservableStack<ViewGenerator>
    
    override public var title: String {
        get {
            return "BleScanDemo"
        }
    }
    public var devices: MutableObservableProperty<Dictionary<String, BleScanResult>>
    
    override public func generate(dependency: ViewDependency) -> View {
        var xml = BleScanDemoXml()
        var view = xml.setup(dependency)
        xml.button.onClick(captureWeak(xml.button){ (button) in 
            Ble.scan(viewDependency: dependency).subscribeBy{ (result) in 
                self.devices.value = self.devices.value + (result.info.id, result)
            }.until(button.removed)
        })
        xml.host.onClick{ () in 
            self.stack.push(BleHostDemoVG())
        }
        xml.items.bind(data: devices.map{ (it) in 
            it.values.sortedByDescending{ (it) in 
                it.rssi
            }
        }, defaultValue: BleScanResult(BleDeviceInfo("", ""), -1000), makeView: { (obs) in 
            var cellXml = ComponentBleDeviceXml()
            var cellView = cellXml.setup(dependency)
            cellXml.deviceName.bindString(obs.map{ (it) in 
                it.info.name ?? "No Name"
            })
            cellXml.deviceId.bindString(obs.map{ (it) in 
                it.info.id
            })
            cellXml.rssi.bindString(obs.map{ (it) in 
                it.rssi.toString()
            })
            cellXml.connect.onClick{ () in 
            }
            return cellView
        })
        return view
    }
    override public func generate(_ dependency: ViewDependency) -> View {
        return generate(dependency: dependency)
    }
    
    public init(stack: ObservableStack<ViewGenerator>) {
        self.stack = stack
        let devices: MutableObservableProperty<Dictionary<String, BleScanResult>> = StandardObservableProperty([:])
        self.devices = devices
        super.init()
    }
    convenience public init(_ stack: ObservableStack<ViewGenerator>) {
        self.init(stack: stack)
    }
}
 
